### PR TITLE
Elementwise and and or

### LINKED ISSUE #25: elementwise_binary doesn't return correct results
I'm using the core.elementwise_binary to set values in a tensor. But it doesn't return the expected result. Unfortunately I do not really understand the code of the function otherwise I would just have made a pull request. Here is a minimal example to reproduce the bug:

```
import sparse as sp
import operator
import numpy as np

size = 3
dimension = 3

coords1 = np.ndarray(shape = 0, dtype=bool)
data1 = np.ndarray(shape = 0, dtype=bool)
tensor1 = sp.COO(coords=coords1, data=data1, shape=((size,) * dimension))
print('emptyTensor')
print(tensor1)
print(tensor1.todense())

coords2 = np.asarray(a=[[0], [1], [2]])
data2 = np.asarray(a=[True], dtype=bool)
tensor2 = sp.COO(coords=coords2, data=data2, shape=((size,) * dimension))
print('\noneElementTensor')
print(tensor2)
print(tensor2.todense())

resultTensor = tensor1.elemwise_binary(operator.or_, tensor2)
print('\nresultTensor = emptyTensor | oneElementTensor')
print(resultTensor)
print(resultTensor.todense())
```

The `resultTensor` should be equal to `tensor2`(`True` at `tensor2[1,2,3]`, all other fields of tensor2 `False`) but it is all fields `False`. 
Additionally I would expect the `dtype` of `resultTensor` to be `bool` because both input `COO`s `tensor1`and `tensor2` are `dtype=bool` and `operator.or_` is generally expected to be a `bool`×`bool`→`bool` operation.

Output:
```
emptyTensor
<COO: shape=(3, 3, 3), dtype=bool, nnz=0, sorted=False, duplicates=True>
[[[False False False]
  [False False False]
  [False False False]]

 [[False False False]
  [False False False]
  [False False False]]

 [[False False False]
  [False False False]
  [False False False]]]

oneElementTensor
<COO: shape=(3, 3, 3), dtype=bool, nnz=1, sorted=False, duplicates=True>
[[[False False False]
  [False False  True]
  [False False False]]

 [[False False False]
  [False False False]
  [False False False]]

 [[False False False]
  [False False False]
  [False False False]]]

resultTensor = emptyTensor | oneElementTensor
<COO: shape=(3, 3, 3), dtype=int64, nnz=0, sorted=False, duplicates=False>
[[[0 0 0]
  [0 0 0]
  [0 0 0]]

 [[0 0 0]
  [0 0 0]
  [0 0 0]]

 [[0 0 0]
  [0 0 0]
  [0 0 0]]]
```

### PR DESCRIPTION
**Pull request by bigerl**
This resolves #25. 
Additionally it introduce the element-wise binary operators AND `&`, OR `|` and XOR `^` on any COO holding Integral values (that includes especially long, int and bool).

### CODE REVIEW

File: sparse/core.py [hunk 1]
```diff
@@ -611,34 +619,46 @@ def elemwise_binary(self, func, other, *args, **kwargs):
...
+            other_coord = tuple(other_coords[j])
+            if self_coord == other_coord:
+                matched_self[i] = True
+                matched_other[j] = True
+                i += 1
+                j += 1
+            elif self_coord < other_coord:
+                i += 1
+            else:
+                j += 1
```
Comment (mrocklin) [inline sparse/core.py]: What is the performance impact of this?  I would expect iterating through a numpy array and creating many tuples to be somewhat expensive.

File: sparse/tests/test_core.py [hunk 2]
```diff
@@ -201,6 +208,23 @@ def test_elemwise_binary_empty():
...
+def test_elemwise_binary_OR():
+    size = 3
+    dimension = 3
+    shape = ((size,) * dimension)
+
+    x = np.zeros(shape=shape,dtype=bool)
+    xx = COO.from_numpy(x)
+
+    y = np.zeros(shape=shape,dtype=bool)
+    y[0,1,2] = True
```
Comment (mrocklin) [inline sparse/tests/test_core.py]: It might be wise to test numeric values as well.  The `|` and `&` operators in NumPy perform bitwise operations.

File: sparse/core.py [hunk 3]
```diff
@@ -611,34 +619,46 @@ def elemwise_binary(self, func, other, *args, **kwargs):
...
 
         # Concatenate matches and mismatches
         data = np.concatenate([func(self_data[matched_self],
                                     other_data[matched_other],
                                     *args, **kwargs),
-                               func(self_data[unmatched_self], 0,
+                               func(self_data[unmatched_self], False,
                                     *args, **kwargs),
-                               func(0, other_data[unmatched_other],
+                               func(False, other_data[unmatched_other],
```
Comment (mrocklin) [inline sparse/core.py]: Why these changes?

Comment (bigerl) [inline sparse/core.py]: The result of a `operator.*` function has the data type of the input with the largest domain. So if you put `0` instead of `False` then `operator.and_(0, False)`will return 0 of data type `int` resulting in an COO with `dtype=int64`instead of the expected `dtype=bool`

File: sparse/core.py (hunk 1 above)
Comment (bigerl) [inline sparse/core.py]: Surly it is not optimal in terms of performance. But the performance should be about the same as running `np.rec.fromarrays(self.coords)` plus `np.rec.fromarrays(other.coords)`. So in the worst case it may affects the performance by about a constant factor 2. 
Generally: his was more a fast approach to fix #25. This part of the code is for my own project quite performance critical. So I will rewrite this section about next week with focused on asymptotic runtime and memory consumption.

File: sparse/tests/test_core.py (hunk 2 above)
Comment (bigerl) [inline sparse/tests/test_core.py]: So does `^`. You're right, I'll just patch that.

File: sparse/core.py (hunk 1 above)
Comment (mrocklin) [inline sparse/core.py]: > But the performance should be about the same as running np.rec.fromarrays(self.coords) plus np.rec.fromarrays(other.coords). So in the worst case it may affects the performance by about a constant factor 2.

How did you come to this conclusion?  Have you compared?  Generally I would expect Python code that generated many tuples to be much slower than NumPy's C code.

File: sparse/tests/test_core.py [hunk 4]
```diff
@@ -392,6 +416,66 @@ def test_scalar_exponentiation():
...
+
+
+def test_int_xor():
+    x = random_x((2, 3, 4), dtype=int)
+    xx = COO.from_numpy(x)
+
+    y = random_x((2, 3, 4), dtype=int)
+    yy = COO.from_numpy(y)
+
+    assert_eq(x ^ y, xx ^ yy)
```
Comment (mrocklin) [inline sparse/tests/test_core.py]: It looks like these tests can be merged using `pytest.mark.parametrize('op', [operator.xor, operator.or_, operator.and_, ...])`

File: sparse/core.py (hunk 1 above)
Comment (mrocklin) [reply sparse/core.py]: This sounds like an asymptotic argument.  Have you actually benchmarked and tested the speed difference here?

What is this change supposed to do?  Is it necessary to handle the elemwise and and or operators?

Comment (bigerl) [reply sparse/core.py]: Of course it is an asymptotic argument and I didn't benchmark it. Sorry, I didn't have time for that then.

Anyways the change from line 634 to 649 is not to handle elemwise AND and OR operations but to handle any elemwise operations correctly.

If you take a look at the example from #25.

It takes two tensors - one empty and one with a True at (0, 1, 2) - and calculates the elemwise OR. The result is an empty tensor. But it should be a tensor with an True or 1 at (0, 1, 2).

You get the same result if you take equivalent int or float valued input tensors and ADD as operation. Result is again an empty (all 0) tensor. So there is definitely a bug in the coords matching between the two input tensors.

I know this is not the most beautiful solution but at least it returns correct results for now.

File: sparse/tests/test_core.py (hunk 4 above)
Comment (bigerl) [reply sparse/tests/test_core.py]: I'll have a look at it next week.

File: sparse/core.py (hunk 3 above)
Comment (hameerabbasi) [reply sparse/core.py]: Technically, I think we should create two zero-elements as follows:

```zero_self = np.zeros((), shape=(), dtype=self.dtype)[()]
zero_other = np.zeros((), shape=(), dtype=other.dtype)[()]```

And use these. I would take into account any casts done by the function correctly. Using `False` or `0` may still be fragile.

**Comment by hameerabbasi** [comment]
@bigerl Is it okay to close this since it was addressed in #38?
